types/type-provider.d.ts

```diff
@@ -8,3 +16,84 @@ export interface FastifyTypeProviderDefault extends FastifyTypeProvider {
+ResolveRequestHeaders<TypeProvider, SchemaCompiler, RouteGeneric>,
+ResolveRequestBody<TypeProvider, SchemaCompiler, RouteGeneric>
+>
+
+// -----------------------------------------------------------------------------------------------
+// FastifyReplyType
+// -----------------------------------------------------------------------------------------------
+
+// Tests if the user has specified a generic argument for Reply
+type UseReplyFromRouteGeneric<RouteGeneric extends RouteGenericInterface> = keyof RouteGeneric['Reply'] extends never ? false : true
+
+// Tests if the user has specified a response schema.
+type UseReplyFromSchemaCompiler<SchemaCompiler extends FastifySchema> = keyof SchemaCompiler['response'] extends never ? false : true
+
+// Resolves the Reply type from the generic argument
+type ResolveReplyFromRouteGeneric<RouteGeneric extends RouteGenericInterface> = RouteGeneric['Reply']
+
+// Resolves the Reply type by taking a union of response status codes
+type ResolveReplyFromSchemaCompiler<TypeProvider extends FastifyTypeProvider, SchemaCompiler extends FastifySchema> = {
+  [K in keyof SchemaCompiler['response']]: CallTypeProvider<TypeProvider, SchemaCompiler['response'][K]>
+} extends infer Result ? Result[keyof Result] : unknown
+
+// The target reply type. This type is inferenced on fastify 'replies' via generic argument assignment
+export type FastifyReplyType<Reply = unknown> = Reply
+
+// Resolves the Reply type either via generic argument or from response schema. This type uses a different
+// resolution strategy to Requests where the Reply will infer a union of each status code type specified
+// by the user. The Reply can be explicitly overriden by users providing a generic Reply type on the route.
+export type ResolveFastifyReplyType<TypeProvider extends FastifyTypeProvider, SchemaCompiler extends FastifySchema, RouteGeneric extends RouteGenericInterface> = FastifyReplyType<
+UseReplyFromRouteGeneric<RouteGeneric> extends true ? ResolveReplyFromRouteGeneric<RouteGeneric> :
+  UseReplyFromSchemaCompiler<SchemaCompiler> extends true ? ResolveReplyFromSchemaCompiler<TypeProvider, SchemaCompiler> :
+    unknown
+>
+
+// -----------------------------------------------------------------------------------------------
+// FastifyReplyReturnType
+// -----------------------------------------------------------------------------------------------
+
+// The target reply return type. This type is inferenced on fastify 'routes' via generic argument assignment
+export type ResolveFastifyReplyReturnType<
+  TypeProvider extends FastifyTypeProvider,
+  SchemaCompiler extends FastifySchema,
+  RouteGeneric extends RouteGenericInterface,
+> = ResolveFastifyReplyType<
+TypeProvider,
+SchemaCompiler,
+RouteGeneric
+> extends infer Return ?
+  (void | Promise<Return | void>)
+// review: support both async and sync return types
+// (Promise<Return> | Return | Promise<void> | void)
```

Comment:
was this missed? will anything be done about it?

we thought to get rid of some unnecessary `async` in our code, but then any `return reply.send()` was flagged as a type error. Having to wrap in `Promise.resolve(reply.send())` (or moving the `return` to another line) isn't quite as ergonomical 😅

Reply:
Can you create a new issue?

Reply:
For sure! #3923